tools/workspace.py

```
@@ -0,0 +1,180 @@
+    url = f"{rawls_host}/api/workspaces/{billing_project_name}/{workspace_name}"
+    token = auth.get_gcp_token()
+
+    headers = auth.build_auth_headers(token)
+    workspace_response = session.get(url=url, headers=headers)
+
+    return workspace_response
+
+
+def delete_workspace(workspace_name: str, billing_project_name: str):
+    """
+    Deletes a workspace from the billing project.
+    :param workspace_name:
+    :param billing_project_name:
+    :return:
+    """
+    token = auth.get_gcp_token()
+    headers = auth.build_auth_headers(token)
+
+    session = get_session_with_retry()
+    workspace_response = get_workspace_by_name(
+        workspace_name, billing_project_name, session
+    )
+    if workspace_response.status_code == 404:
+        logging.info(
+            f"Workspace {billing_project_name}/{workspace_name} is gone, skipping deletion."
+        )
+        return
+
+    workspace = workspace_response.json()
+    workspace_status = workspace["workspace"]["state"]
+    if workspace_status in ["Ready", "DeleteFailed"]:
+        logging.info(
+            f"Workspace {billing_project_name}/{workspace_name} status is {workspace_status}, starting deletion"
+        )
+        rawls_host = Configuration.get_config()["rawls_host"]
+        request = (
+            f"{rawls_host}/api/workspaces/v2/{billing_project_name}/{workspace_name}"
+        )
+        response = session.delete(url=request, headers=headers)
+        response.raise_for_status()
+    elif workspace_status in ["Deleting"]:
+        logging.info(
+            f"Workspace {billing_project_name}/{workspace_name} is already being deleted, beginning poll..."
+        )
+
+    def deletion_poller():
+        raw_status = get_workspace_by_name(
+            workspace_name, billing_project_name, session
+        )
+        if raw_status.status_code == 404:
+            return True, None
+
+        if "workspace" not in raw_status.json():
+            logging.info(raw_status.json())
```

Should you raise an exception instead of letting the rest of the function proceed?

---

Good idea, added https://github.com/broadinstitute/dsp-terra-workspaces-scripts/pull/13/commits/0633b0086b5fbd6109337459a06576261916be32